### PR TITLE
Change sharedtria dof distribution to match distributed tria

### PR DESCRIPTION
**Pull request by tcclevenger**
This was discussed here: https://github.com/dealii/dealii/issues/3078#issuecomment-341225860
Talking with @tjhei,  we felt it best to change from "coin flip" assignment to "smallest subdomain" assignment for those active dofs on the interface between two subdomains of a `p::shared::T` so that we match the distribution in `p::distributed::T` (this is also how we chose to distribute mg dofs for `p::shared::T`).
Note:
- There are 4 different tests (16 total runs)  that we expect to fail with this change. This is due to the fact that each of these test output `n_locally_owned_dofs`. This value will obviously change when we change the way in which we break ties along the interface.
- I have added two tests, i) testing against a p4est mesh for accuracy against `p::distributed::T` and ii) creating an adaptive mesh and outputting the  owned dofs for each subdomain and dof numbering for each cell.
- I changed some filenames from tests I wrote for a previous pull request. Just a wording change.

### CODE REVIEW

File: tests/sharedtria/dof_05.mpirun=3.output [hunk 1]
```diff
@@ -0,0 +1,14 @@
+
+{[0,20]}
```
Comment (tjhei) [inline tests/sharedtria/dof_05.mpirun=3.output]: you need to put p4est into the filename

**Comment by agrayver** [comment]
For the record: `DoFTools::subdomain_wise` (or, more specifically, `DoFTools::get_subdomain_association`) also uses the coin flip strategy and this patch does not change it (either intentionally or not).

**Comment by tjhei** [comment]
> (either intentionally or not).

Yes. I was thinking about this yesterday. Those functions can go away. I will make a PR that deprecates them.

**Comment by bangerth** [comment]
This would deserve a changelog entry in the "incompatibilities" section.

**Comment by tcclevenger** [comment]
I'm a bit curious as to why the checks actually passed considering I have not yet fixed the test which fail to pass on my computer due to the change. Looking at the build details it doesn't appear to have run any sharedtria tests. I just still go back and update these, correct?

**Comment by bangerth** [comment]
Yes, you need to update these anyway. The tester only runs a subset of the sequential tests because running the entire test suite would take too long.

**Comment by davydden** [comment]
> I have not yet fixed the test which fail to pass on my computer due to the change

let us know when you check locally that all relevant tests pass with this PR. I put a WIP tag for now...

**Comment by tcclevenger** [comment]
All tests for sharedtria now pass.

**Comment by tcclevenger** [comment]
I've added the changelog entry and all relevant test have passed. Let me know if there are any other changes needed, if not everything should be good to go. Thanks!

**Comment by jppelteret** [comment]
@tjhei @davydden @kronbichler Can this be merged?

**Comment by davydden** [comment]
I have no further comments, ok to merge as far as I am concerned.

**Comment by tjhei** [comment]
I will go ahead and press "merge" then.

**Pull request merged**